### PR TITLE
Add policy logging blog post

### PR DESCRIPTION
**Pull request by ereslibre**
Fixes: https://github.com/kubewarden/kubewarden.io/issues/22

### CODE REVIEW

File: content/blog/2021/11/deep-dive-into-policy-logging.md [hunk 1]
```diff
@@ -0,0 +1,218 @@
+---
+title: Deep Dive into policy logging
+authors:
+    - Rafael Fernández López
+date: 2021-11-15
+---
+
+Policies are regular programs. As such they have the need to log often times. In general, we are
```
Comment (viccuad) [inline content/blog/2021/11/deep-dive-into-policy-logging.md:8]: Nit:

```suggestion
Policies are regular programs. As such they often have the need to log timestamps. In general, we are
```

File: content/blog/2021/11/deep-dive-into-policy-logging.md [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+used to make our programs log into standard output (stdout) and standard error (stderr) outputs.
+
+However, policies run in a confined WebAssembly environment. For this mechanism to work as usual
+Kubewarden would need to set up the runtime environment in a way that stdout and stderr file
+descriptors are set up so that the policy can write to them, and upon completion, Kubewarden can
```
Comment (viccuad) [inline content/blog/2021/11/deep-dive-into-policy-logging.md:13]: Nit:

I would reword it to:
 "Kubewarden would need to set up the runtime environment in a way that the policy can write to stdout and stderr file
descriptors, and upon completion, Kubewarden can.."

(the 2 chained "set up" and the long subordinate sentence seems difficult to parse for me)

File: content/blog/2021/11/deep-dive-into-policy-logging.md [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+check them -- or stream them as they pop up.
+
+Although the previous approach was our very first initial implementation in Kubewarden, it's no
+longer implemented that way.
+
+Given we are providing SDK's for supported languages we can do a bit better. As we have described in
+previous posts and in our documentation, currently Kubewarden uses waPC for allowing
```
Comment (viccuad) [inline content/blog/2021/11/deep-dive-into-policy-logging.md:20]: Nit:

```suggestion
previous posts and in our documentation, Kubewarden currently uses waPC for allowing
```

It seems that using "currently" as an introducing word, in the position that it is right now, needs a comma after. But then it's too staccato. Hence, the swapping to using it as pure adverb so it flows faster.

File: content/blog/2021/11/deep-dive-into-policy-logging.md [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+
+The waPC interface used for logging is a contract, because once you have built a policy, it should
+be possible to run it in future Kubewarden versions. In this sense, we have preferred to keep this
+contract behind the SDK of your preferred language, so you don't have to deal with the details of
+how logging is implemented in Kubewarden. You just have to use your preferred logging library for
+the language of your choice.
+
+Let's look into how to take advantage of logging with Kubewarden in specific languages!
+
+## Policy Authors
```
Comment (viccuad) [inline content/blog/2021/11/deep-dive-into-policy-logging.md:35]: Nit:

```suggestion
## For policy Authors
```
(And the same in "For cluster administrators")

Just so it feels warmer.

**Review by viccuad — APPROVED** [review]
Nice! Added some nitpicking optional changes, yet it's already ready for me.

File: content/blog/2021/11/deep-dive-into-policy-logging.md [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+descriptors are set up so that the policy can write to them, and upon completion, Kubewarden can
+check them -- or stream them as they pop up.
+
+Although the previous approach was our very first initial implementation in Kubewarden, it's no
+longer implemented that way.
+
+Given we are providing SDK's for supported languages we can do a bit better. As we have described in
+previous posts and in our documentation, Kubewarden uses waPC for allowing intercommunication
+between the guest (the policy) and the host (Kubewarden -- the `policy-server` or `kwctl` if we are
+running policies manually).
```
Comment (flavio) [inline content/blog/2021/11/deep-dive-into-policy-logging.md:22]: This will need some small reworking once the previous sentence is dropped ofc.

File: content/blog/2021/11/deep-dive-into-policy-logging.md [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+
+Kubewarden has defined  a contract between policies (guests) and the host (Kubewarden) for
+performing [policy settings
+validation](https://docs.kubewarden.io/writing-policies/spec/02-settings.html), [policy
+validation](https://docs.kubewarden.io/writing-policies/spec/03-validating-policies.html), [policy
+mutation](https://docs.kubewarden.io/writing-policies/spec/04-mutating-policies.html) and logging.
```
Comment (flavio) [inline content/blog/2021/11/deep-dive-into-policy-logging.md:28]: Slightly OT, I feel like we should extend the docs to cover the interface of logging too... Can you file an issue for that?

File: content/blog/2021/11/deep-dive-into-policy-logging.md (hunk 5 above)
Comment (flavio) [reply content/blog/2021/11/deep-dive-into-policy-logging.md:22]: Also, please add a link to waPC website. It's actually pretty nice: https://wapc.io/

File: content/blog/2021/11/deep-dive-into-policy-logging.md [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+```go
+logger.InfoWithFields("validating object", func(e onelog.Entry) {
+    e.String("name", gjson.GetBytes(payload, "request.object.metadata.name").String())
+	e.String("namespace", gjson.GetBytes(payload, "request.object.metadata.namespace").String())
+})
+```
+
+What happens under the covers is that our Go SDK sends [every log event to the `kubewarden` host
+through
+waPC](https://github.com/kubewarden/policy-sdk-go/blob/95c4b93adf6830862b85f953bc09d31de7490872/log_writer_wasi.go#L15).
```
Comment (flavio) [inline content/blog/2021/11/deep-dive-into-policy-logging.md:85]: I have mixed feelings about these sections. I think we can drop them for the sake of reducing the word count on the whole article (and keep the reader focused and engaged)

File: content/blog/2021/11/deep-dive-into-policy-logging.md [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+info!(
+    LOG_DRAIN,
+    "rejecting resource";
+    "resource_name" => &resource_name
+);
+```
+
+As happens with the Go SDK implementation, our Rust implementation of the `slog` drain sends [this
+logging events to the host by using
+waPC](https://github.com/kubewarden/policy-sdk-rust/blob/642d13b07053be6455c0b5f49df0e87ff6f022c5/src/logging/drain.rs#L57-L60).
```
Comment (flavio) [inline content/blog/2021/11/deep-dive-into-policy-logging.md:126]: Same here

File: content/blog/2021/11/deep-dive-into-policy-logging.md [hunk 9]
```diff
@@ -0,0 +1,221 @@
...
+
+You can read more about slog [here](https://github.com/slog-rs/slog).
+
+### Swift
+
+We will be looking at the [Swift policy
+template](https://github.com/kubewarden/swift-policy-template) for this example.
+
+As happens with Go and Rust's SDK's, the Swift SDK is instrumented to use Swift's `LogHandler`, so
+our policy only has to initialize it. In our `Sources/Policy/main.swift` file:
```
Comment (flavio) [inline content/blog/2021/11/deep-dive-into-policy-logging.md:136]: We're using the popular [swift-log](https://github.com/apple/swift-log). Can you write that and drop a link to the GitHub repo?

File: content/blog/2021/11/deep-dive-into-policy-logging.md [hunk 10]
```diff
@@ -0,0 +1,221 @@
...
+            "some-key": "some-value",
+        ])
+
+    // ...
+}
+```
+
+Following the same strategy as the Go and Rust SDK's, the Swift SDK is able to [push log events to
+the host through
+waPC](https://github.com/kubewarden/policy-sdk-swift/blob/59cc979fa9994f1653d3d3e3dd3072188b2c0a18/Sources/kubewardenSdk/logger.swift#L75-L80).
```
Comment (flavio) [inline content/blog/2021/11/deep-dive-into-policy-logging.md:165]: Same here

File: content/blog/2021/11/deep-dive-into-policy-logging.md [hunk 11]
```diff
@@ -0,0 +1,221 @@
...
+
+## Cluster Administrators
+
+Being able to log from within a policy is half of the story. Then, we have to be able to read and
+potentially collect these logs.
+
+As we have seen, Kubewarden policies support structured logging that is then forwarded to the
+component running the policy. Usually, this is `kwctl` if you are executing the policy in a manual
+fashion, or the `policy-server` if the policy is being ran in a Kubernetes environment.
```
Comment (flavio) [inline content/blog/2021/11/deep-dive-into-policy-logging.md:174]:
```suggestion
fashion, or `policy-server` if the policy is being ran in a Kubernetes environment.
```

File: content/blog/2021/11/deep-dive-into-policy-logging.md [hunk 12]
```diff
@@ -0,0 +1,221 @@
...
+
+Both `kwctl` and the  `policy-server` use the `tracing` crate to produce log events, either the
```
Comment (flavio) [inline content/blog/2021/11/deep-dive-into-policy-logging.md:176]:
```suggestion
Both `kwctl` and  `policy-server` use the [`tracing`](https://github.com/tokio-rs/tracing) crate to produce log events, either the
```

File: content/blog/2021/11/deep-dive-into-policy-logging.md [hunk 13]
```diff
@@ -0,0 +1,221 @@
...
+The `kwctl` CLI tool takes a very straightforward approach to logging from policies: it will print
+them to the standard error file descriptor.
+
+### `policy-server`
+
+The `policy-server` supports [different log
+formats](https://github.com/kubewarden/policy-server/blob/d615bfe7fdf1fe6001e655360fe85f0db2194410/src/cli.rs#L174-L206):
+`json`, `text` and `otlp`.
+
+`otlp`? I hear you ask. It stands for OpenTelemetry Protocol. We will look into that in a bit.
```
Comment (flavio) [inline content/blog/2021/11/deep-dive-into-policy-logging.md:191]:
```suggestion
`otlp`? I hear you ask. It stands for [OpenTelemetry](https://opentelemetry.io/) Protocol. We will look into that in a bit.
```

File: content/blog/2021/11/deep-dive-into-policy-logging.md [hunk 14]
```diff
@@ -0,0 +1,221 @@
...
+
+If the `policy-server` is run with the `--log-fmt` argument set to `json` or `text`, the output will
+be printed to the standard error file descriptor in JSON or plain text formats.
```
Comment (flavio) [inline content/blog/2021/11/deep-dive-into-policy-logging.md:194]: Maybe add:

> These messages then be read using `kubectl logs <policy-server-pod>`

File: content/blog/2021/11/deep-dive-into-policy-logging.md [hunk 15]
```diff
@@ -0,0 +1,221 @@
...
+
+If `--log-fmt` is set to `otlp`, the `policy-server` will use OpenTelemetry to report logs and
+traces.
+
+## OpenTelemetry
+
+Kubewarden is instrumented with OpenTelemetry, so it's possible for the `policy-server` to send
```
Comment (flavio) [inline content/blog/2021/11/deep-dive-into-policy-logging.md:201]:
```suggestion
Kubewarden is instrumented with [OpenTelemetry](https://opentelemetry.io/), so it's possible for the `policy-server` to send
```

File: content/blog/2021/11/deep-dive-into-policy-logging.md [hunk 16]
```diff
@@ -0,0 +1,221 @@
...
+  enabled: True
+  tracing:
+    jaeger:
+      endpoint: "all-in-one-collector.jaeger.svc.cluster.local:14250"
+```
+
+This functionality closes the gap on logging/tracing, given the freedom that the OpenTelemetry
+collector provides to us in terms of flexibility of what to do with this logs and traces.
+
+But this is a big enough topic on its own worth a future blog post. Stay logged!
```
Comment (flavio) [inline content/blog/2021/11/deep-dive-into-policy-logging.md:221]: Before closing, I would add a link to the section of our docs where we explain how to setup tracing.

**Review by flavio — CHANGES_REQUESTED** [review]
Great writing, thanks a lot for having done that.

I left some comments around

File: content/blog/2021/11/deep-dive-into-policy-logging.md (hunk 1 above)
Comment (ereslibre) [reply content/blog/2021/11/deep-dive-into-policy-logging.md:8]: I will change it to emit messages, but honestly that might mean so many different things. Emit messages to whom, where? While the act of logging is a well known action that comprises certain behaviors (e.g. write a message next to a timestamp, possibly with a component tag, possibly with metadata in a certain format...).

File: content/blog/2021/11/deep-dive-into-policy-logging.md (hunk 6 above)
Comment (ereslibre) [reply content/blog/2021/11/deep-dive-into-policy-logging.md:28]: Filed https://github.com/kubewarden/docs/issues/69

File: content/blog/2021/11/deep-dive-into-policy-logging.md (hunk 9 above)
Comment (ereslibre) [reply content/blog/2021/11/deep-dive-into-policy-logging.md:136]: The link was some lines after, I can add another link here.

File: content/blog/2021/11/deep-dive-into-policy-logging.md (hunk 7 above)
Comment (ereslibre) [reply content/blog/2021/11/deep-dive-into-policy-logging.md:85]: I would prefer to keep them, since if you are interested in a language it's one click away, it's not like we are embedding more code making the post huge.

Also, rendered does not seem so long to me.

Comment (ereslibre) [reply content/blog/2021/11/deep-dive-into-policy-logging.md:85]: I also think it adds value because if you are interested as a reader you can see that basically they approach waPC in the same way.

File: content/blog/2021/11/deep-dive-into-policy-logging.md (hunk 15 above)
Comment (ereslibre) [reply content/blog/2021/11/deep-dive-into-policy-logging.md:201]: We added a link on the first occurence of the term. Should I keep adding links to it?

**Comment by ereslibre** [comment]
Updated the PR with a separate commit so it's easier to review. I will squash prior to merge.

File: content/blog/2021/11/deep-dive-into-policy-logging.md (hunk 1 above)
Comment (ereslibre) [reply content/blog/2021/11/deep-dive-into-policy-logging.md]: Changed it to log information.

File: content/blog/2021/11/deep-dive-into-policy-logging.md (hunk 7 above)
Comment (viccuad) [reply content/blog/2021/11/deep-dive-into-policy-logging.md:83]: I'm happy with them, helped me learn the specifics.

**Review by jvanz — APPROVED** [review]
LGTM! :D

File: content/blog/2021/11/deep-dive-into-policy-logging.md (hunk 7 above)
Comment (flavio) [reply content/blog/2021/11/deep-dive-into-policy-logging.md:83]: Fine with me :)

File: content/blog/2021/11/deep-dive-into-policy-logging.md (hunk 15 above)
Comment (flavio) [reply content/blog/2021/11/deep-dive-into-policy-logging.md:202]: No we shouldn't. Sorry, I missed the first link

**Review by flavio — APPROVED** [review]
LGTM, please reach out to Rober to get his opinion about the whole article

**Comment by robertsirc** [comment]
Looks good to me. I would cross post it in kubewarden.io and in commuinty.suse.com.

**Comment by ereslibre** [comment]
Live at:

- https://www.kubewarden.io/blog/2021/11/deep-dive-into-policy-logging/
- https://community.suse.com/posts/kubewarden-deep-dive-into-policy-logging
